content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,511 @@
+boolean functions. They take some boolean arguments and return a boolean, e.g.
+
+      f: B³ → B
+    (X,Y,Z) ↦ ((¬X∧Y)∨Z)∧(X∨¬Y)
+
+[It turns out](https://en.wikipedia.org/wiki/Functional_completeness) that *every*
+boolean function with boolean parameters can be written as a logical formula -
+in particular, every such function has a CNF and a DNF.
+
+If we want to do that for other sets (apart from `B^n`, that is), we first have
+to be able to translate it into a set of boolean variables. As we will see,
+this can be easily done in the case we are interested in.
+
+All of this allows us to talk about what we mean by “computing a set is hard”.
+With this equivalency, checking if a given object is in a set becomes the
+*decision* problem of SAT - deciding if `x∈S` is equivalent to deciding if
+`f[S](x)`evaluates to `true`. And checking if a set is empty becomes the
+*solution* problem of SAT, as it is equivalent to verifying that `f[S]` is not
+satisfiable.
+
+With this, let us look at the specific sets we are interested in.
+
+## Basic interfaces as type sets
+
+Interfaces in Go have always described sets of types. For example, the interface
+
+```go
+type S interface {
+    X()
+    Y()
+    Z()
+}
+```
+
+is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.
+
+We also have a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):
+
+
+```go
+type S interface { X() }
+type T interface { Y() }
+type U interface {
+    S
+    T
+}
+```
+
+This expresses `T=S∩T` as an interface. We can also view the predicate “has a
+method `X()`” as a boolean variable and think of this as the formula `X∧Y`.
+
+Surprisingly, there is also a simple form of negation. It happens implicitly,
+because a type can not have two methods with the same name, but different
+types. So, implicitly, if a type has a method `X()`, it does *not* have a
+method `X() int`, for example[^5]:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+```
+
+This meant that we can express interfaces which could never be satisfied by
+any type (i.e. which describe an empty type set):
+
+```go
+interface{ X; NotX }
+```
+
+[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).
+
+So we had a language to build boolean formulas, where every formula had the
+form
+
+    X∧Y∧¬Z
+
+Checking if a type implements an interface means checking if it is in the set
+described by that interfaces. Which means checking if the corresponding formula
+is satisfied.
+
+And checking if an interface is invalid meant checking if the set it describes
+is empty. Which meant checking if the formula is *satisfiable*. This was easy,
+though, as those formulas are in DNF. They only have a single term and that
+term is a conjunction of variables or their negations. And as we noted above,
+solving SAT on a formula given in DNF is easy.
```

Comment:
These three paragraphs switch between present & past tense pretty fluidly and it's confusing.

```
So we HAVE a language to build boolean formulas, where every formula HAS the
form

    X∧Y∧¬Z

Checking if a type implements an interface means checking if it is in the set
described by that interfaces. Which means checking if the corresponding formula
is satisfied.

And checking if an interface is invalid meanS checking if the set it describes
is empty. Which meanS checking if the formula is *satisfiable*. This IS easy,
though, as those formulas are in DNF. They only have a single term and that
term is a conjunction of variables or their negations. And as we noted above,
solving SAT on a formula given in DNF is easy.
```

Reply:
I blame Roger.
Fixed it.